### PR TITLE
adds changes for implementation of ISL ranges

### PR DESCRIPTION
**Pull request by desaikd**
*Issue #10*

*Description of changes:*
This PR works on creating ISL ranges.

*Changes:*
* adds changes for ISL `Range`
  *  `RangeBoundaryValue`: represents a range boundary value (i.e. min, max or a value in terms of `RangeBoundaryValueType`)
  * `RangeBoundaryValueType`: represents a range boundary value which can be a `Decimal`/`Integer`/...
  * `RangBoundaryType`: defines exclusivity of the range boundary
* adds a bugfix for unordered types comparison (a new `PartialEq` implementation for `IslTypeImpl`)
* adds unit tests related to ranges

*Test:*
Adds unit tests for ISL `Range`s

### CODE REVIEW

File: src/isl/isl_type.rs [hunk 1]
```diff
@@ -119,3 +119,18 @@ impl IslTypeImpl {
         Ok(IslTypeImpl::new(type_name, constraints))
     }
 }
+
+// OwnedStruct doesn't preserve field order hence the PartialEq won't work properly for unordered constraints
+// Related issue: https://github.com/amzn/ion-rust/issues/200
+impl PartialEq for IslTypeImpl {
```
Comment (popematt) [inline src/isl/isl_type.rs:125]: non-blocking question—can you explain why we're using `PartialEq` instead of `Eq` for all of the types in this library? As far as I can tell, the only difference is reflexivity, and I can't think of any case for ISL where `a != a`. (I know that `nan` exists and `nan != nan`, but in the context of ISL, something like `valid_values: [ X ]` has reflexive equality even if `X` is `nan`.)

File: src/isl/util.rs [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+///                       | range::[ <EXCLUSIVITY><RANGE_TYPE>, max ]
+/// Grammar: <RANGE_TYPE> ::= <DECIMAL>
+///                | <FLOAT>
+///                | <INT>
+///                | <NUMBER>
+///                | <TIMESTAMP>
+///                | <TIMESTAMP_PRECISION_VALUE>
+/// For more information on [Range]: <https://amzn.github.io/ion-schema/docs/spec.html#constraints>
+#[derive(Debug, Clone, PartialEq)]
+pub enum Range {
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/isl/util.rs:22]: Yes, I will create an issue for it.

File: src/isl/mod.rs [hunk 3]
```diff
@@ -223,4 +228,100 @@ mod isl_tests {
...
+    fn owned_struct_to_range(range1: IonSchemaResult<Range>, range2: IonSchemaResult<Range>) {
+        // determine that both the ranges are created with no errors
+        assert_eq!(range1.is_ok(), true);
+        assert_eq!(range2.is_ok(), true);
+
+        // assert if both the ranges are same
+        assert_eq!(range1.unwrap(), range2.unwrap());
+    }
+
+    #[rstest(
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/isl/mod.rs:299]: I can add the tests for mismatch types but for `range::` annotation I feel it should be added wherever range is actually used (e.g. `byte_length: range::[2,5]`).

File: src/isl/util.rs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                value.as_timestamp().unwrap().to_owned(),
+                range_boundary_type,
+            )),
+            _ => return invalid_schema_error("Unsupported range type specified"),
+        }
+    }
+}
+
+/// Represents the range boundary types in terms of exclusivity (i.e. inclusive or exclusive)
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/isl/util.rs:282]: I had added this because of the comparison to `start > end` validation but as it depends on timestamp ordering which is not yet implemented, I can remove it for now.

File: src/isl/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (popematt) [reply src/isl/mod.rs:299]: Okay

File: src/isl/util.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (popematt) [reply src/isl/util.rs:282]: If it's something that we for sure need later, don't remove it now just because I don't understand it.

The `start > end` validation relies on comparing `RangeBoundaryValue`s or possibly `RangeBoundaryValueType`s. I understand why the `RangeBoundaryValueType` should have `PartialOrd`, but I don't understand why `RangeBoundaryType` has any ordering at all. Is `Inclusive` greater than `Exclusive` or is it the other way around?

File: src/isl/util.rs [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+        let range_type = Range::validate_range_boundaries(&start, &end)?;
+        match range_type {
+            RangeBoundaryValueType::Decimal(_) => Ok(Range::Decimal(start, end)),
+            RangeBoundaryValueType::Float(_) => Ok(Range::Float(start, end)),
+            RangeBoundaryValueType::Integer(_) => Ok(Range::Integer(start, end)),
+            RangeBoundaryValueType::IntegerNonNegative(_) => {
+                Ok(Range::IntegerNonNegative(start, end))
+            }
+            RangeBoundaryValueType::Timestamp(_) => Ok(Range::Timestamp(start, end)),
+        }
```
Comment (zslayton) [inline src/isl/util.rs:92]: If you make a consolidated `range` function, this would just be `range(start, end)`.

File: src/isl/util.rs [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+    }
+
+    // helper method to which validates a non negative integer range boundary value
+    fn validate_non_negative_integer_range_boundary_value(
+        value: &AnyInt,
+    ) -> IonSchemaResult<usize> {
+        match value.as_i64() {
+            Some(v) => {
+                if v > 0 {
```
Comment (zslayton) [inline src/isl/util.rs:101]: I think this should be `>=`?

File: src/isl/util.rs [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+                Ok(Range::IntegerNonNegative(start, end))
+            }
+            RangeBoundaryValueType::Timestamp(_) => Ok(Range::Timestamp(start, end)),
+        }
+    }
+
+    // helper method to which validates a non negative integer range boundary value
+    fn validate_non_negative_integer_range_boundary_value(
+        value: &AnyInt,
+    ) -> IonSchemaResult<usize> {
```
Comment (zslayton) [inline src/isl/util.rs:98]: The size of `usize` varies by platform and is usually used for addresses in memory or offsets. On some platforms, this would end up being a 4-byte integer.

Should we just return the same `AnyInt` so we can continue to support values of arbitrary size?

File: src/isl/util.rs [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+    }
+    pub fn int_non_negative_value(value: usize, range_boundary_type: RangeBoundaryType) -> Self {
+        RangeBoundaryValue::Value(
+            RangeBoundaryValueType::IntegerNonNegative(value),
+            range_boundary_type,
+        )
+    }
+    pub fn float_value(value: f64, range_boundary_type: RangeBoundaryType) -> Self {
+        RangeBoundaryValue::Value(RangeBoundaryValueType::Float(value), range_boundary_type)
+    }
```
Comment (zslayton) [inline src/isl/util.rs:214]: Similar TODO thought for these: I think we could have a consolidated `value` method?

File: src/isl/util.rs [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+
+    pub fn timestamp_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Timestamp(start, end))
+    }
+
+    pub fn from_ion_element(value: &OwnedElement, is_non_negative: bool) -> IonSchemaResult<Range> {
```
Comment (zslayton) [inline src/isl/util.rs:68]: The `is_non_negative` flag feels a bit unfortunate to me. It's only relevant in the integer case to distinguish between `Integer` and `IntegerNonNegative` but has to be specified no matter which type is expected. It also has to be passed through to each of the nested calls to other `from_ion_element` implementations.

Looking at [the grammar](https://amzn.github.io/ion-schema/docs/spec.html#grammar), it seems like `RANGE_TYPE` doesn't include a `NonNegativeInteger` variant: 
```
<RANGE_TYPE> ::= <DECIMAL>
               | <FLOAT>
               | <INT>
               | <NUMBER>
               | <TIMESTAMP>
               | <TIMESTAMP_PRECISION_VALUE>
```

Instead, it [creates a type](https://amzn.github.io/ion-schema/docs/spec.html#byte_length-constraint) to model this constraint:

```
type::{
  name: range_int_non_negative,
  type: list,
  annotations: required::[range],
  ordered_elements: [
    { one_of: [ int_non_negative, { valid_values: [min] } ] },
    { one_of: [ int_non_negative, { valid_values: [max] } ] },
  ],
  container_length: 2,
}
```

I'm inclined to say we should follow suit here. We would remove the `NonNegativeInteger` variant from the `RangeBoundaryValueType` and model this as a type instead. `from_ion_element` would then be able to infer everything it needs from the Ion data itself.

File: src/isl/util.rs (hunk 7 above)
Comment (desaikd) [reply src/isl/util.rs:98]: If I understand this correctly, you are suggesting we should represent both `integer` and `non_negative_integers` as `AnyInt`. I think that should be okay as we already annotate range boundaries with it's type.

File: src/isl/util.rs [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+    Timestamp(RangeBoundaryValue, RangeBoundaryValue),
+}
+
+impl Range {
+    pub fn decimal_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Decimal(start, end))
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/isl/util.rs:36]: This is verified inside `validate_range_boundaries` when both the boundaries are `Value`s.

File: src/isl/util.rs (hunk 9 above)
Comment (desaikd) [reply src/isl/util.rs:68]: I have created an issue for this to change it after I have ordered_elements implemented as a constraint. (Reference issue: https://github.com/amzn/ion-schema-rust/issues/46)

File: src/isl/util.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (desaikd) [reply src/isl/util.rs:22]: Reference issue: https://github.com/amzn/ion-schema-rust/issues/47

File: src/isl/util.rs [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+        Ok(Range::IntegerNonNegative(start, end))
+    }
+
+    pub fn timestamp_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Timestamp(start, end))
+    }
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/isl/util.rs:66]: Reference issue: https://github.com/amzn/ion-schema-rust/issues/48

File: src/isl/util.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (desaikd) [reply src/isl/util.rs]: I have also added unit test for this as popematt@ suggested in his comments.

File: src/isl/util.rs [hunk 12]
```diff
@@ -79,33 +91,18 @@ impl Range {
...
-        }
+        // validate both range boundary values and returns created `Range`
+        Range::range(start, end)
     }
 
     // helper method to which validates a non negative integer range boundary value
     fn validate_non_negative_integer_range_boundary_value(
         value: &AnyInt,
-    ) -> IonSchemaResult<usize> {
+    ) -> IonSchemaResult<&AnyInt> {
```
Comment (zslayton) [inline src/isl/util.rs:101]: Does this still need to return a value?

**Review by zslayton — APPROVED** [review]

File: src/isl/util.rs (hunk 12 above)
Comment (desaikd) [reply src/isl/util.rs:101]: No. I can remove it and just return `IonSchemaResult<()>` instead.
